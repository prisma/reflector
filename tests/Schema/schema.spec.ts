import { PrismaUtils } from '~/index'

describe('parseDatasourceOrThrow', () => {
  describe('success', () => {
    describe('provider type input is accepted, but normalized upon return', () => {
      it('postgresql becomes postgres', () => {
        const result1 = PrismaUtils.Schema.parseDatasourceOrThrow(`
          datasource db {
            provider = "postgres"
            url = env("...")
          }
        `)
        const result2 = PrismaUtils.Schema.parseDatasourceOrThrow(`
          datasource db {
            provider = "postgresql"
            url = env("...")
          }
        `)
        expect(result1.provider).toEqual(PrismaUtils.Schema.ProviderTypeNormalized._def.values.postgres)
        expect(result1.provider).toEqual(result2.provider)
      })
    })
    describe('ignores comments', () => {
      test('a whole datasource block', () => {
        expect(
          PrismaUtils.Schema.parseDatasourceOrThrow(`
            //datasource db_old {
            //  url = env("...old")
            //}
            // datasource db_old_2 {
            datasource db {
              url = env("...")
              provider = "postgres"
            }
          `)
        ).toMatchSnapshot()
      })
      test('single field lines', () => {
        expect(
          PrismaUtils.Schema.parseDatasourceOrThrow(`
            datasource db {
              // url = env("bad1")
              url = env("...")
              ///  url = "bad2"
            //blah url = "bad3"
                ///blah url = "bad4"
              provider = "postgres"
            }
          `)
        ).toMatchSnapshot()
      })
    })

    test('finds inline connection string', () => {
      expect(
        PrismaUtils.Schema.parseDatasourceOrThrow(`
          datasource db {
            url = "..."
              provider = "postgres"
          }
        `)
      ).toMatchSnapshot()
    })
    describe('environment variable reference', () => {
      test('finds when nicely formatted', () => {
        expect(
          PrismaUtils.Schema.parseDatasourceOrThrow(`
            datasource db {
              url = env("FOOBAR")
              provider = "postgres"
            }
          `)
        ).toMatchSnapshot()
      })

      test('finds when white space before or after quotes', () => {
        expect(
          PrismaUtils.Schema.parseDatasourceOrThrow(`
            datasource db {
              url = env(  "FOOBAR"  )
              provider = "postgres"
            }
          `)
        ).toMatchSnapshot()
      })
    })
  })
  describe('thrown errors', () => {
    test('multiple data source blocks causes an error', () => {
      expect(() =>
        PrismaUtils.Schema.parseDatasourceOrThrow(`
          datasource db1 {
            url = "..."
          }

          datasource db2 {
            url = "..."
          }
        `)
      ).toThrowErrorMatchingSnapshot()
    })

    test('Datasource block missing url field causes an error', () => {
      expect(() =>
        PrismaUtils.Schema.parseDatasourceOrThrow(`
          datasource db {}
        `)
      ).toThrowErrorMatchingSnapshot()
    })
    test('Missing a provider field causes an error', () => {
      expect(() =>
        PrismaUtils.Schema.parseDatasourceOrThrow(`
          datasource db {
            url = "..."
          }
        `)
      ).toThrowErrorMatchingSnapshot()
    })
    test('An invalid provider field causes an error', () => {
      expect(() =>
        PrismaUtils.Schema.parseDatasourceOrThrow(`
          datasource db {
            url = "..."
            provider = "sqlitee"
          }
        `)
      ).toThrowErrorMatchingSnapshot()
    })
  })
})
