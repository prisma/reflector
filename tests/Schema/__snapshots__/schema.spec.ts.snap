// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseDatasourceOrThrow success environment variable reference finds when nicely formatted 1`] = `
Object {
  "_tag": "ParsedDatasourceUrlEnvironmentVariable",
  "environmentVariableName": "FOOBAR",
  "name": "db",
  "provider": "postgres",
  "sources": Object {
    "url": "env(\\"FOOBAR\\")",
  },
}
`;

exports[`parseDatasourceOrThrow success environment variable reference finds when white space before or after quotes 1`] = `
Object {
  "_tag": "ParsedDatasourceUrlEnvironmentVariable",
  "environmentVariableName": "FOOBAR",
  "name": "db",
  "provider": "postgres",
  "sources": Object {
    "url": "env(  \\"FOOBAR\\"  )",
  },
}
`;

exports[`parseDatasourceOrThrow success finds inline connection string 1`] = `
Object {
  "_tag": "ParsedDatasourceUrlInline",
  "connectionString": "...",
  "name": "db",
  "provider": "postgres",
  "sources": Object {
    "url": "\\"...\\"",
  },
}
`;

exports[`parseDatasourceOrThrow success ignores comments a whole datasource block 1`] = `
Object {
  "_tag": "ParsedDatasourceUrlEnvironmentVariable",
  "environmentVariableName": "...",
  "name": "db",
  "provider": "postgres",
  "sources": Object {
    "url": "env(\\"...\\")",
  },
}
`;

exports[`parseDatasourceOrThrow success ignores comments single field lines 1`] = `
Object {
  "_tag": "ParsedDatasourceUrlEnvironmentVariable",
  "environmentVariableName": "...",
  "name": "db",
  "provider": "postgres",
  "sources": Object {
    "url": "env(\\"...\\")",
  },
}
`;

exports[`parseDatasourceOrThrow thrown errors An invalid provider field causes an error 1`] = `"Failed to parse datasource: No valid provder property set."`;

exports[`parseDatasourceOrThrow thrown errors Datasource block missing url field causes an error 1`] = `"Failed to parse datasource: No valid provder property set."`;

exports[`parseDatasourceOrThrow thrown errors Missing a provider field causes an error 1`] = `"Failed to parse datasource: No valid provder property set."`;

exports[`parseDatasourceOrThrow thrown errors Missing a url field causes an error 1`] = `
"Failed to parse datasource: No url property found:

Pattern:
/^\\\\s*url\\\\s*=\\\\s*([^\\\\n]+)\\\\s*$/gm

Code:
datasource db {
            provider = \\"sqlite\\"
          }"
`;

exports[`parseDatasourceOrThrow thrown errors multiple data source blocks causes an error 1`] = `"Failed to parse datasource: Multiple datasource blocks found."`;
